src/hotspot/share/cds/filemap.cpp

```diff
@@ -1916,6 +1936,43 @@ char* FileMapInfo::map_bitmap_region() {
   return bitmap_base;
 }
 
+class SharedDataRelocationTask : public ArchiveWorkerTask {
+private:
+  BitMapView* const _rw_bm;
+  BitMapView* const _ro_bm;
+  SharedDataRelocator* const _rw_reloc;
+  SharedDataRelocator* const _ro_reloc;
+
+public:
+  SharedDataRelocationTask(BitMapView* rw_bm, BitMapView* ro_bm, SharedDataRelocator* rw_reloc, SharedDataRelocator* ro_reloc) :
+                           ArchiveWorkerTask("Shared Data Relocation"),
+                           _rw_bm(rw_bm), _ro_bm(ro_bm), _rw_reloc(rw_reloc), _ro_reloc(ro_reloc) {}
+
+  void work(int chunk, int max_chunks) override {
+    BitMapView* bm;
+    SharedDataRelocator* reloc;
+    int bitmap_chunk;
+
+    // Pick up the slice from one of the bitmaps.
+    int chunks_per_bitmap = max_chunks / 2;
+    if (chunk < chunks_per_bitmap) {
+      bm = _rw_bm;
```

Comment:
How is it guaranteed that the two regions will each have half the chunks?

Reply:
I admit this is confusing, and likely implicitly relies on max_chunks being divisible by 2. The intent here was to "map" chunks [0..max/2) of one bitmap, and [max/2, max) to another bitmap. The actual "pivot" of max/2 should not have mattered much, since once we get into actual slice computation, we look back to the actual size of the bitmap. I think it could be _significantly more straight-forward_ if we just did "a piece of one bitmap, and a piece of another bitmap" as the unit of work. See new commit?

Reply:
Yes, this is definitely simpler and clearer.